File: src/main/java/LeetCode49.java
import java.util.*;

public class LeetCode49 {
    public static void main(String[] strings) {
        String[] strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
        System.out.println(groupAnagrams(strs));
    }
    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> group = new HashMap<>();

        for (String str : strs) {
            char[] temp = str.toCharArray();
            Arrays.sort(temp);
            String key = String.valueOf(temp);
            group.putIfAbsent(key, new ArrayList<>());
            group.get(key).add(str);
        }

        return new ArrayList<>(group.values());
    }
}

File: src/main/kotlin/LeetCode49.kt
fun main(args: Array<String>) {
    val instance = LeetCode49()
    val input = arrayOf("eat", "tea", "tan", "ate", "nat", "bat")
    val res = instance.groupAnagrams(input)
    println(res)
}

class LeetCode49 {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val res = hashMapOf<String, MutableList<String>>()
        strs.forEach {
            val key = String(it.toCharArray().sortedArray())
            res[key]?.let { list ->
                list.add(it)
            } ?: run {
                res[key] = mutableListOf(it)
            }
        }
        return res.values.toList()
    }
}